### DISCUSSION TITLE
Smaller, Faster-starting Container Images With jlink and AppCDS - Gunnar Morling

### DISCUSSION THREAD

**Discussion by blog-comment-importer** [gunnarmorling/discussions.morling.dev#280] [Announcements]
https://www.morling.dev/blog/smaller-faster-starting-container-images-with-jlink-and-appcds/

**Comment by blog-comment-importer** [comment]
_**Utilisateur** wrote at 2021-01-16T00:14:00Z (comment imported from Disqus):_

<p>Really cool article: thanks for the detailed process for the generation of the image of the modular application.<br>Found you from Java Annotated Monthly <a href="https://blog.jetbrains.com/idea/2021/01/java-annotated-monthly-january-2021/" rel="nofollow noopener" title="https://blog.jetbrains.com/idea/2021/01/java-annotated-monthly-january-2021/">https://blog.jetbrains.com/...</a></p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-02-02T17:49:07Z (comment imported from Disqus):_

<p>Excellent, thanks! Very happy to hear the post was useful to you.</p>

**Comment by blog-comment-importer** [comment]
_**rafael grisotto** wrote at 2021-02-02T16:28:05Z (comment imported from Disqus):_

<p>Great article. Did you see improvement in the container memory usage with the custom JRE?</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-02-02T17:50:29Z (comment imported from Disqus):_

<p>I didn't check memory consumption (heap, or RSS), but I'd expect it to not make any substantial difference. Classes which are not loaded won't consume memory, so using a custom image via jlink shouldn't make a difference here.</p>

**Reply by simonis** [reply]
You are right in that you won't get huge memory savings by using AppCDS. However, if you start several Docker instances from the same image, the read-only parts of the CDS/AppCDS archives can be shared between the instances and the savings will increase with each new instance you start.
My (a little bit outdated) [talk about AppCDS](https://youtu.be/-u4ZK8hcy44) has [a slide on this](https://simonis.github.io/JPrime2018/CDS/cds.xhtml#/9).
There's also a nice (and more recent) [article on this topic by Vladimir Plizga](https://medium.com/@toparvion/appcds-for-spring-boot-applications-first-contact).

**Reply by gunnarmorling** [reply]
@simonis, yeah, good point. I'm wondering though how relevant that is in practice; when looking at Kubernetes, you'll hardly ever run more than one Java in a pod, so you won't benefit from that. Now if there was a way to share CDS archives across pod boundaries on one K8s worker node(assuming it can be safely done), than this would be very interesting.

**Reply by simonis** [reply]
I'm not very familiar with Kubernetes/Pods. In the end it bails down to wether the Pods / Container images get loaded by the same kernel from the same file system. In the end this is simple COW sharing done by the kernel automatically through page cache. I.e. if the CDS archives in different PODs/containers have the same inode number, they will be subject to sharing.
